### PR TITLE
New Instrument: Siglent SPDxxxxX series Power Supplies

### PR DESCRIPTION
**Pull request by AidenDawn**
Add new instrument, SPD1168X power supply with base and channel class for the SPD power supply family. #718 would mean the channel implementation here could rely on a different base class but this works for now. The only unimplemented function is currently the ability to set the timer steps of the PSU, this requires setting a list of values in the same command which I couldn't find an example for. Any pointers to one would be great so I can add that last property

### CODE REVIEW

**Comment by bmoneke** [comment]
> The only unimplemented function is currently the ability to set the timer steps of the PSU, this requires setting a list of values in the same command

Do I understand it correctly, that you have to write a list (instead of a single value) to the device?

Two solutions:
1. You can use either a validator or a set_process parameter for a pymeasure property, because the value you give it, is processed by both: `value = set_process(validator(value, values))`. The set_process could turn your list into a string appropriate for the device and your set_command could look something like `"TIMeSTeP %s"` introducing the string you previously made.
```
time_step = Instrument.setting("TIMeSTeP %s", "The time steps", set_process=lambda v: str(v))
# you can give another method (static or outside the class) as set_process as well.
inst.time_step = [1, 2, 3, 10]
```
2. You can always use normal python properties, if need arises (in this case set_process would do the job):
```
@time_step.setter
def time_step(self, values):
    values = ":".join([str(value) for value in values])
    self.write(f"TIMeSTeP {values}")
```

**Comment by AidenDawn** [comment]
Thanks. I'll have a look into that and have a play around to get it working. Just for a bit more context to my question, the command structure is as follows:

**Comment by bmoneke** [comment]
Another question: Why did you choose to make a "base Instrument"? Are there more instruments, which just differ in the channel number? Maybe you could read the instrument version (via id) and create that much channels (see active technolgies AWG401x).

We recently added [protocol tests](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html#protocol-tests), which do not require a connected instrument. Please provide at least some tests, as they allow us to maintain the code (e.g. changing the channel implementation) without having access to the instrument proper.

**Comment by AidenDawn** [comment]
> Another question: Why did you choose to make a "base Instrument"? Are there more instruments, which just differ in the channel number? Maybe you could read the instrument version (via id) and create that much channels (see active technolgies AWG401x).

The family of instruments includes some 3 channel and some single channel instruments which can all have different voltage and current output properties. I was trying to leave the option open in the future to include these other instruments. I prefer how the channels have been created in that AWG. I'll make some edits and update

> We recently added [protocol tests](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html#protocol-tests), which do not require a connected instrument. Please provide at least some tests, as they allow us to maintain the code (e.g. changing the channel implementation) without having access to the instrument proper.

Will do, looks like Ive got some more work to do ;)

File: pymeasure/instruments/siglenttechnologies/siglent_spd1168x.py
Comment (bmoneke) [inline pymeasure/instruments/siglenttechnologies/siglent_spd1168x.py:43]: Docstrings should be an imperative, what the method should do:
```suggestion
        """ Ensure that the voltage is turned to zero
        and disable the output. """
```

File: pymeasure/instruments/siglenttechnologies/siglent_spd1168x.py
Comment (bmoneke) [inline pymeasure/instruments/siglenttechnologies/siglent_spd1168x.py:58]: Did you intend to write a test?

For this, we have a test file folder (see documentation). Tests with devices have a different file from tests without a device.

Some of the present tests with devices show, how to do it. They have, e.g. a parameter area, where you can specify the device address.

Comment (AidenDawn) [reply pymeasure/instruments/siglenttechnologies/siglent_spd1168x.py:58]: I'm just in the process of writing the test now that I've finished the implementation. Should come through as a final commit if there are no issues

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (bmoneke) [inline pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:151]: Please move the __init__s to the top of the classes, to find them more easily.

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (bmoneke) [inline pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:173]: Why do you return the instrument instance?

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (bmoneke) [inline pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:185]: Does this work as expected?

The property creators work at class level, therefore, all channels will be "CH1".

What you can do, is to define a (**not** f-string) `"MEAS:VOLT? CH{channel}"` and in `SPDChannel.write` you do `self.instrument.write(command.format(channel=self.channel))`. This way, you replace at instance level the channel name.

Comment (AidenDawn) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:185]: Thanks for this one. I was slightly worried that was what I had done here. Good to have some confirmation. Changes are in the works

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (AidenDawn) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:173]: You're right. These returns aren't necessary. They were left over from a copy of another instrument and I missed them

**Comment by AidenDawn** [comment]
This should now be ready pending a final review of the edits made

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (AidenDawn) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:98]: Thanks, I can see why that makes sense. I've definitely seen this style included in other instruments before so I inlcuded it here, but Im removing it now

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (bmoneke) [inline pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:229]: Does the docstring match the values?

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (bmoneke) [inline pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:257]: You say, you pass on kwargs, but you don't do it.
```suggestion
        """ Reads a set of values from the instrument through the adapter,
        passing on any key-word arguments.
        """
        return self.instrument.values(cmd.format(channel=self.channel), **kwargs)
```

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (bmoneke) [inline pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:79]: I've rarely seen this type annotation. For a getter (measurement), most of the time, the type does not matter, as you get the variable in the correct type. It's only important for the setter and there not often. Most of the time it is obvious (boolean, string, float, or int).
Additionally, you can write it in the text, if the type might be ambibous, e.g.:
```suggestion
        """Read the error code and information of the instrument as a string.
```

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (bmoneke) [inline pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:129]: Is the type really a class or just an index? According to "values", I'd say its an integer.

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (bmoneke) [inline pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:135]: I'd add a comment, that instrument implementors have to change the values according to the number of channels. At a first glance, it looks stupid to have a single valid value.

**Review by bmoneke — COMMENTED** [review]
A few more comments from my side. Some are only suggestions on the docstrings, and some are minor inconsistencies. Thanks for the patience with my comments.

Now the maintainers (@bilderbuchi @msmttchr ) have to look over the code. They will make some more suggestions and requests.

Thank you again for your contribution.

**Comment by AidenDawn** [comment]
> A few more comments from my side. Some are only suggestions on the docstrings, and some are minor inconsistencies. Thanks for the patience with my comments.
> 
> Now the maintainers (@bilderbuchi @msmttchr ) have to look over the code. They will make some more suggestions and requests.
> 
> Thank you again for your contribution.

Thanks for the comments, it's a good chance to see where I can improve what Im sending out/missing throughout this process

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (AidenDawn) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:79]: In the case of this instrument, the error code (returned as an integer in a string format) is followed by a plain-text description of the error which has occurred. In the default case, the string "0 No Error" is the result which I agree is an unusual response. You're right that your suggestion is probably more readable as a sentence but this version of formatting should show up very clearly in the docs. The docs page for temptronic_base was my example for this

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (AidenDawn) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:257]: Good spot

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (bilderbuchi) [inline pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:273]: If you can wait for some time until #718 lands, the whole channel handling will become much easier, with less need to define these methods.

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (bilderbuchi) [inline pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:294]: Why does this duplicate what `output` does? 
Is that the only function that needs that channel selection machinery? If yes, would using the channels `.output` alone even work correctly?
Why are you returning `self` here?

File: pymeasure/instruments/siglenttechnologies/siglent_spd1168x.py
Comment (bilderbuchi) [inline pymeasure/instruments/siglenttechnologies/siglent_spd1168x.py:58]: This looks like it's leftover from testing?

File: tests/instruments/siglenttechnologies/test_siglent_spd1168x.py
Comment (bilderbuchi) [inline tests/instruments/siglenttechnologies/test_siglent_spd1168x.py:53]: You give it 10, it truncates to 8, but then returns 0.5? That seems confusing? :smile:

File: tests/instruments/test_all_instruments.py
Comment (bilderbuchi) [inline tests/instruments/test_all_instruments.py:57]: What happens if you don't do that? From what I can see, `SPDBase` does _not_ need communication in its init?

**Review by bilderbuchi — CHANGES_REQUESTED** [review]
Thanks @bmoneke for the preparatory work!
Some remarks from my side. If you can wait a little for #718 to land, this contribution can be streamlined a little more.

**Comment by bilderbuchi** [comment]
> Thanks. I'll have a look into that and have a play around to get it working. Just for a bit more context to my question, the command structure is as follows:
> 
> TIMEr Command format TIMEr: SET <SOURce>, <secnum>, <\volt>, <curr>, <time> <SOURce>:={CH1}; <secnum>;=1 to 5; Description Set timing parameters of specified channel Example TIMEr: SET CH1, 2, 3, 0.5, 2

For this, as it's a pretty complicated thing, I recommend implementing a method `configure_timer`. You can add arguments with clear names for all the needed values, and it sounds like an action to begin with. Not everything is a great fit for properties :-)

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (AidenDawn) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:111]: Youre right, the name needs changing. I think disable_local_interface is probably even better

File: pymeasure/instruments/siglenttechnologies/__init__.py
Comment (AidenDawn) [reply pymeasure/instruments/siglenttechnologies/__init__.py:25]: I've been using temptronic instruments as the template here. I think if you were implementing an instrument in the family that hadn't been specifically included already then there is scope for the exposure of the base class to be useful. But your point is also completely valid. Happy to go either way

File: tests/instruments/siglenttechnologies/test_siglent_spd1168x.py
Comment (AidenDawn) [reply tests/instruments/siglenttechnologies/test_siglent_spd1168x.py:53]: Yes, that's odd

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (AidenDawn) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:217]: For some reason, when I try and and build the docs without this space I end up with:
`C:\Users\dschmeer\Projects\pymeasure\pymeasure\instruments\siglenttechnologies\siglent_spdbase.py:docstring of pymeasure.instruments.siglenttechnologies.siglent_spdbase.SPDChannel:1: WARNING: Field list ends without a blank line; unexpected unindent.`
It did this in a few places in SPDBase, is there a known reason for this? If not Ill just leave the newline in

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (AidenDawn) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:249]: We'll go with enable_timer, particularly given I'm writing set_timer as a method too

File: tests/instruments/test_all_instruments.py
Comment (AidenDawn) [reply tests/instruments/test_all_instruments.py:57]: This is for the same reason as `ATSBase`. Neither of these two base classes provide an instrument name, which consistently causes the __init__ of `Instrument` to fail. This is not needed for the child classes because the name is defined in there. The other option I tried was to set a base name (something like Siglent SPDxxxxX instrument Base Class) but that caused further errors in the child classes with multiple `name` definitions

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (AidenDawn) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:294]: I've changed the output part to remove property and replace with an `enable_output` method. hopefully this will make more sense overall

File: tests/instruments/test_all_instruments.py
Comment (bmoneke) [reply tests/instruments/test_all_instruments.py:57]: @bilderbuchi the comment above the list states the reason for the exception: these instruments require input arguments (in this case `name`).

I know, that the skip message is not appropriate, but I did not add a special check just for `Instrument` class.

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (AidenDawn) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:217]: I've done a bit more testing, it seems like there is a sphinx whitespace bug going on here. I have two options which work, one is the one included above. The other option is to replace the line in the docstring as:
`  :type : float
   """,`. Personally my preference will be to keep the newline because it looks tidier

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (bilderbuchi) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py]: `disable_local_interface` would [suggest a method](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html#your-instrument-s-user-interface), not a property, though, as `disable` is a verb.

File: pymeasure/instruments/siglenttechnologies/__init__.py
Comment (bilderbuchi) [reply pymeasure/instruments/siglenttechnologies/__init__.py:25]: Hm, good point. If you're implementing a new instrument, I think you should preferably directly import from the`siglent_spdbase` subpackage, and the main package namespace would be for completed instruments. 
We don't have a formulated policy for this, yet.  @pymeasure/pymeasure-maintainers thoughts?

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (bilderbuchi) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:116]: Don't "resolve" conversation where you have a question. :-)

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (bilderbuchi) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py]: Hm, could be a Sphinx/ReST quirk or bug (couldn't find a report in the sphinx repo with a quick search), resolve it as you want. Great that you checked doc generation locally!

File: tests/instruments/test_all_instruments.py
Comment (bilderbuchi) [reply tests/instruments/test_all_instruments.py:57]: My only excuse is that it must have been too late to notice that comment :-P

What if you add `kwargs.setdefault('name', 'Siglent SPDxxxxX instrument Base Class')` in `SPDBase`'s init, just before you call `super()__init__`? This should enable using it in tests, but still result in the correct names for the child classes?

File: pymeasure/instruments/siglenttechnologies/siglent_spd1168x.py
Comment (bilderbuchi) [inline pymeasure/instruments/siglenttechnologies/siglent_spd1168x.py:60]: also add a call to `super().shutdown()` at the end, so you get the rest of the cleanup, too.

Comment (bilderbuchi) [reply pymeasure/instruments/siglenttechnologies/siglent_spd1168x.py:60]: Also, since you seem to duplicate this in the other child class, maybe that should move up to `SPCBase`?

File: pymeasure/instruments/siglenttechnologies/siglent_spd1305x.py
Comment (bilderbuchi) [inline pymeasure/instruments/siglenttechnologies/siglent_spd1305x.py:66]: as above.

File: pymeasure/instruments/siglenttechnologies/siglent_spd1168x.py
Comment (bilderbuchi) [inline pymeasure/instruments/siglenttechnologies/siglent_spd1168x.py:43]: I'm confused, why is this property now duplicated in the two child classes?

Comment (bilderbuchi) [reply pymeasure/instruments/siglenttechnologies/siglent_spd1168x.py:43]: Also, from the name (having a verb) this should be a method `enable_4W_mode(enabled):`

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (bilderbuchi) [inline pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:101]: A verb in the name implies an "action", which is most properly a method, not a property, see https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html#your-instrument-s-user-interface.

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (bilderbuchi) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:116]: I'm not sure I fully understand the feature distribution here, but maybe have an really slim class `SPDSingleChannel(SPDBase)` if you want to isolate the single-channel-model-only features?

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (bilderbuchi) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py]: or did that work out in the end?

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (bilderbuchi) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:116]: You only need this for `enable_output`, then, and only in setting? Might be easier to just add a second write call to that method?

**Comment by bmoneke** [comment]
I agree, that "base instruments" (for some manufacturers etc.) should not be in the main name space. As a user, you won't need them, but they will pop up with autocompletion and might make it more difficult to find the actual instrument. 

Only implementing an instrument, you need them, and then you can use the slightly longer name. 

Am 13. November 2022 11:58:25 MEZ schrieb Christoph Buchner ***@***.***>:
***@***.*** commented on this pull request.

File: pymeasure/instruments/siglenttechnologies/__init__.py
Comment (msmttchr) [reply pymeasure/instruments/siglenttechnologies/__init__.py:25]: I agree. My preference would be to not include it in the main package namespace.

File: tests/instruments/test_all_instruments.py
Comment (AidenDawn) [reply tests/instruments/test_all_instruments.py]: This works

File: pymeasure/instruments/siglenttechnologies/siglent_spdbase.py
Comment (bilderbuchi) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:234]: I agree, this seems ready pending the channel changes. 
I would say wait until #718 lands, merge a fresh master into your branch, then rework. 
Alternatively, @bmoneke do you want this in master now, pull into #718 and rework this instrument within #718?

Comment (bmoneke) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:234]: There are many instruments with channels, which I do not rework in #718.
We can rework them step by step, as they will continue to work as they are (just the code base is not as nice).

Let's merge this work as is, in order to close the PR

Comment (bilderbuchi) [reply pymeasure/instruments/siglenttechnologies/siglent_spdbase.py:234]: Sounds good! I opened #746 for tracking this.

**Pull request merged**
